routing/bicycle_directions.cpp

```diff
@@ -0,0 +1,185 @@
+   double m_routeLength;
+};
+
+ftypes::HighwayClass GetHighwayClass(FeatureID const & featureId, Index const & index)
+{
+  ftypes::HighwayClass highWayClass = ftypes::HighwayClass::Undefined;
+  MwmSet::MwmId const & mwmId = featureId.m_mwmId;
+  uint32_t const featureIndex = featureId.m_index;
+
+  FeatureType ft;
+  Index::FeaturesLoaderGuard loader(index, mwmId);
+  loader.GetFeatureByIndex(featureIndex, ft);
+  ft.ParseGeometry(FeatureType::BEST_GEOMETRY);
+  highWayClass = ftypes::GetHighwayClass(ft);
+  ASSERT_NOT_EQUAL(highWayClass, ftypes::HighwayClass::Error, ());
+  ASSERT_NOT_EQUAL(highWayClass, ftypes::HighwayClass::Undefined, ());
+  return highWayClass;
+}
+}  // namespace
+
+namespace routing
+{
+BicycleDirectionsEngine::BicycleDirectionsEngine(Index const & index) : m_index(index)
+{
+}
+
+void BicycleDirectionsEngine::Generate(IRoadGraph const & graph, vector<Junction> const & path,
+                                       Route::TTimes & times,
+                                       Route::TTurns & turnsDir,
+                                       vector<m2::PointD> & routeGeometry,
+                                       my::Cancellable const & cancellable)
+{
+  LOG(LINFO, ("BicycleDirectionsEngine::Generate"));
+  CHECK_GREATER(path.size(), 1, ());
+  times.clear();
+  turnsDir.clear();
+  routeGeometry.clear();
+  m_adjacentEdges.clear();
+  m_pathSegments.clear();
+
+  CalculateTimes(graph, path, times);
+
+  auto emptyPathWorkaround = [&]()
+  {
+    turnsDir.emplace_back(path.size() - 1, turns::TurnDirection::ReachedYourDestination);
+    this->m_adjacentEdges[0] = {{}, 1}; // There's one ingoing edge to the finish.
+  };
+
+  IRoadGraph::TEdgeVector routeEdges;
+  if (!ReconstructPath(graph, path, routeEdges, cancellable))
+  {
+    LOG(LDEBUG, ("Couldn't reconstruct path"));
+    emptyPathWorkaround();
+    return;
+  }
+  if (routeEdges.empty())
+  {
+    ASSERT(false, ());
+    emptyPathWorkaround();
+    return;
+  }
+
+  // Filling |m_adjacentEdges|.
+  size_t const pathSize = path.size();
+  m_adjacentEdges.insert(make_pair(0, AdjacentEdges({{} /* outgoingEdges */, 0 /* ingoingEdges.size() */})));
+  for (size_t i = 1; i < pathSize; ++i)
+  {
+    Junction const & formerJunction = path[i - 1];
+    Junction const & currentJunction = path[i];
+    IRoadGraph::TEdgeVector outgoingEdges, ingoingEdges;
+    graph.GetOutgoingEdges(currentJunction, outgoingEdges);
+    graph.GetIngoingEdges(currentJunction, ingoingEdges);
+
+    AdjacentEdges adjacentEdges = {{}, ingoingEdges.size()};
+    adjacentEdges.outgoingTurns.reserve(outgoingEdges.size());
+    for (auto const & outgoingEdge : outgoingEdges)
+    {
+      auto const & outgoingFeatureId = outgoingEdge.GetFeatureId();
+      // Checking for if |outgoingEdge| is a fake edge.
+      if (!outgoingFeatureId.m_mwmId.IsAlive())
+        continue;
+      double const angle = turns::PiMinusTwoVectorsAngle(formerJunction.GetPoint(),
```

Comment:
Why do you need an angle between (formerJunction.GetPoint(), currentJunction.GetPoint()) and
(formerJunction.GetPoint(), outgoingEdge.GetEndJunction().GetPoint()) ?
Why not (formerJunction.GetPoint(), currentJunction.GetPoint()) and (currentJunction.GetPoint(), outgoingEdge.GetEndJunction().GetPoint())?